core/node/node_framework/src/resource/unique.rs

```diff
@@ -0,0 +1,35 @@
+use std::sync::{Arc, Mutex};
+
+/// Wrapper for resources that only support one consumer.
+///
+/// Normally, all the resources should support sharing between several tasks,
+/// but there are some cases where a resource should only be consumed by a single task.
+#[derive(Debug)]
+pub struct Unique<T: 'static + Send> {
+    inner: Arc<Mutex<Option<T>>>,
```

Comment:
1: Maybe, what I'm afraid of is that it's not an approach widely used in our codebase, so it may not be intuitive.
2. Again maybe, what I'm afraid of here is that `Unique` is opt-in: you have to explicitly say that you're working with an unique resource on both ends. `take_resource` may be used by accident and cause problems.

I'll think about both points more, and maybe will implement in upcoming PRs.